src/sig_stfl/xmss/sig_stfl_xmss_secret_key_functions.c

```diff
@@ -27,37 +87,65 @@ OQS_STATUS OQS_SECRET_KEY_XMSS_serialize_key(const OQS_SIG_STFL_SECRET_KEY *sk,
 	*sk_buf_ptr = sk_buf;
 	*sk_len = sk->length_secret_key;
 
+	/* Unlock the key if possible */
+	OQS_SECRET_KEY_XMSS_activate_unlock(sk);
+
 	return OQS_SUCCESS;
 }
 
-/* Deserialize XMSS byte string into an XMSS secret key data */
+/* Deserialize XMSS byte string into an XMSS secret key data. */
 OQS_STATUS OQS_SECRET_KEY_XMSS_deserialize_key(OQS_SIG_STFL_SECRET_KEY *sk, const size_t sk_len, const uint8_t *sk_buf, XMSS_UNUSED_ATT void *context) {
 	if (sk == NULL || sk_buf == NULL || (sk_len != sk->length_secret_key)) {
 		return OQS_ERROR;
 	}
 
-	if (sk->secret_key_data != NULL) {
-		OQS_MEM_secure_free(sk->secret_key_data, sk->length_secret_key);
-		sk->secret_key_data = NULL;
+	memcpy(sk->secret_key_data, sk_buf, sk->length_secret_key);
+	sk->context = context;
+
+	return OQS_SUCCESS;
+}
+
+void OQS_SECRET_KEY_XMSS_set_store_cb(OQS_SIG_STFL_SECRET_KEY *sk, secure_store_sk store_cb, void *context) {
+	if (sk == NULL || store_cb == NULL) {
+		return;
 	}
+	sk->secure_store_scrt_key = store_cb;
+	sk->context = context;
+}
 
-	// Assume key data is not present
-	sk->secret_key_data = malloc(sk_len);
-	if (sk->secret_key_data == NULL) {
-		return OQS_ERROR;
+void OQS_SECRET_KEY_XMSS_free(OQS_SIG_STFL_SECRET_KEY *sk) {
+	if (sk == NULL) {
+		return;
 	}
 
-	sk->context = context;
-	memcpy(sk->secret_key_data, sk_buf, sk_len);
+	OQS_MEM_secure_free(sk->secret_key_data, sk->length_secret_key);
+	sk->secret_key_data = NULL;
+}
 
-	return OQS_SUCCESS;
+void OQS_SECRET_KEY_XMSS_activate_lock(OQS_SIG_STFL_SECRET_KEY *sk) {
+	if (sk == NULL) {
+		return;
+	}
+
+	if (sk->is_locked == false) {
```

Comment:
Not I understand this. If the mutex is already locked subsequent lock requests are ignored?

Reply:
Correct. May be it's not the best way, I need to the `is_locked` variable to be thread-safe here. 
Back to the logic: 
I think if we lock the secret key, we would need only one lock. 
So if the key is locked, then the subsequent lock won't need to lock it again until the lock is released.

Reply:
For my understanding ... You added this logic to prevent a crash?
The way the code is currently, another thread would be allowed to access the already locked secret key. This not what we would want.
Secondly, the current thread, holding the lock, should only try to call activate_lock once unless there is some recursive calls using the same key. I would expect the system to handle that case. And if the 'activate_lock' is called by another thread, then we want it blocked.
Thats my understanding on the behavior we want. Your thoughts?

Reply:
> For my understanding ... You added this logic to prevent a crash?

I added this to mainly prevent a hang. 

>  The way the code is currently, another thread would be allowed to access the already locked secret key. This not what we would want.

I agree totally.

> Secondly, the current thread, holding the lock, should only try to call activate_lock once unless there is some recursive calls using the same key. I would expect the system to handle that case. And if the 'activate_lock' is called by another thread, then we want it blocked.

That's right, it's my ultimate goal as well. I'm not sure the current approach is the best approach, it's mainly to avoid the hang. Feel free to edit the code, I have basic knowledge with mutex so I think I'm not doing the right job here.

Reply:
Do you have a backtrace of the actual hang?

Reply:
Hmm, I don't have a backtrace. I remember that if I comment out the `if is_lock == false` check in both activate/deactivate locks then it will hang.

Reply:
Can you use gdb to get the traceback?

Reply:
Can you rerun your test with but comment out the call to "activate_lock" from 	OQS_SECRET_KEY_XMSS_activate_lock(sk); at line #77  in sig_stfl_xmss_secret_key_functions.c

Reply:
<img width="540" alt="image" src="https://github.com/open-quantum-safe/liboqs/assets/106774416/611204a9-c84a-4ca5-9239-2c8d7689dcdc">

I'm having a mac with me right now so it's LLDB.

Reply:
> Can you rerun your test with but comment out the call to "activate_lock" from OQS_SECRET_KEY_XMSS_activate_lock(sk); at line https://github.com/open-quantum-safe/liboqs/pull/77 in sig_stfl_xmss_secret_key_functions.c

I commented out the `activate_lock` and `activate_unlock`, it works normally.

Reply:
Can you rebuild with debug symbols? Maybe CFLAGS="-g -O0"

Reply:
I think it's better if you jump into the current branch and asset the hang. 
After you build, run the command below to reproduce

```
$gdb ./build/tests/test_sig_stfl XMSS-SHA2_10_256 tests/KATs/sig_stfl/xmss/XMSS-SHA2_10_256.rsp
```

Reply:
Looks like we just need to use a recursive mutex.

Reply:
@ducnguyen-sb can you take a look at the tests that are timing out?

Reply:
Yup, I think I know why. Will update asap.

Reply:
Okay, today I have time to analyze deeper this problem. 
Turn out, using RECURSIVE MUTEX increase runtime significantly. 

Benchmark all XMSS parameters in my M1 Mac.

- No thread test: 82 seconds
- With RECURSIVE MUTEX: 120 seconds (50% increase)
- Rearrange function, simple MUTEX; 120 seconds.  (50% increase)

So basically it's time out because of the use of MUTEX. 
I think the solution here is to increase timeout a little bit more.